### PR TITLE
fix: API response-headers compatibility with origin #131

### PR DESCRIPTION
**Pull request by sonbui00**
API response-headers compatibility with the origin
Fix: https://github.com/mccutchen/go-httpbin/issues/141 

```
curl  "http://127.0.0.1:8080/response-headers?Bar=bar1&Bar=bar2&Foo=foo"                   
{
  "Bar": [
    "bar1",
    "bar2"
  ],
  "Content-Length": 104,
  "Content-Type": "application/json; charset=utf-8",
  "Foo": "foo"
}

curl "https://httpbin.org/response-headers?Bar=bar1&Bar=bar2&Foo=foo"
{
  "Bar": [
    "bar1", 
    "bar2"
  ], 
  "Content-Length": "127", 
  "Content-Type": "application/json", 
  "Foo": "foo"
}
```

### CODE REVIEW

File: httpbin/handlers.go [hunk 1]
```diff
@@ -314,15 +314,26 @@ func (h *HTTPBin) Unstable(w http.ResponseWriter, r *http.Request) {
...
+	body := make(map[string]interface{})
 	for k, vs := range args {
 		for _, v := range vs {
 			w.Header().Add(k, v)
 		}
+		if len(vs) == 1 {
+			body[k] = vs[0]
+		} else {
+			body[k] = vs
+		}
```
Comment (mccutchen) [inline httpbin/handlers.go:326]: Please drop this logic.  Per [my comment here](https://github.com/mccutchen/go-httpbin/issues/91#issuecomment-1667003449)

> But, to be clear, go-httpbin will continue to serialize header values (and form values and any other “zero or more” values inherent to HTTP) as JSON arrays for consistency, where original httpbin serializes them as single scalar values, arrays, or comma-separated strings, depending on the endpoint and incoming request parameters.

That might not be very clear.  To restate it, what I mean is that go-httpbin returns all header values as JSON arrays, even if only one header is given.

I know that is not actually backwards-compatible with the original httpbin, but for now I've decided that I prefer the consistency (and implementation simplicity) of this approach and I'm okay with the incompatibility.  (See https://github.com/mccutchen/go-httpbin/issues/37 and https://github.com/mccutchen/go-httpbin/issues/15 for related discussion.)

File: httpbin/handlers.go [hunk 2]
```diff
@@ -314,15 +314,26 @@ func (h *HTTPBin) Unstable(w http.ResponseWriter, r *http.Request) {
 // ResponseHeaders responds with a map of header values
 func (h *HTTPBin) ResponseHeaders(w http.ResponseWriter, r *http.Request) {
 	args := r.URL.Query()
+	body := make(map[string]interface{})
```
Comment (mccutchen) [inline httpbin/handlers.go:317]: With my comment below, I don't think we need anything more than `args` here.

File: httpbin/handlers.go [hunk 3]
```diff
@@ -314,15 +314,26 @@ func (h *HTTPBin) Unstable(w http.ResponseWriter, r *http.Request) {
...
 	if contentType := w.Header().Get("Content-Type"); contentType == "" {
 		w.Header().Set("Content-Type", jsonContentType)
 	}
-	mustMarshalJSON(w, args)
+	body["Content-Type"] = w.Header().Get("Content-Type")
+	body["Content-Length"] = 0
+	buf := &bytes.Buffer{}
+	json.NewEncoder(buf).Encode(body)
+	body["Content-Length"] = buf.Len()
+	mustMarshalJSON(w, body)
```
Comment (mccutchen) [inline httpbin/handlers.go:336]: This is a bit of a strange approach, double-encoding the body just to get the content length.  Please rewrite it to do the encoding once.  The existing [`writeResponse()` helper func](https://github.com/mccutchen/go-httpbin/blob/62ec5553f2401137c24a0cc67f04c5fd278a4e4b/httpbin/helpers.go#L96-L100) might be useful here.

File: httpbin/handlers_test.go [hunk 4]
```diff
@@ -1163,16 +1163,32 @@ func TestResponseHeaders(t *testing.T) {
...
+			Foo           string   `json:"Foo"`
+			Bar           []string `json:"Bar"`
+			ContentType   string   `json:"Content-Type"`
+			ContentLength int      `json:"Content-Length"`
+		}
+
+		result := mustParseResponse[bodyResponse](t, resp)
+
+		assert.ContentType(t, resp, result.ContentType)
+		assert.Equal[bool](t, result.ContentLength > 0, true, "JSON response Content-Length mismatch")
```
Comment (mccutchen) [inline httpbin/handlers_test.go:1177]: I think we can do away with most of this specialization if we simplify the response as requested above.

**Review by mccutchen — COMMENTED** [review]
Thanks for the contribution!  I've left a few comments below that will help simplify these changes, please let me know if they make sense to you.

Also, have you explored what the original httpbin's behavior is in weird circumstances?  Like, what happens with a request like `/response-headers?Content-Length=999999999`?  I think we need some test coverage for at least that edge case, to codify whatever behavior we decide on.

**Comment by mccutchen** [comment]
(Also, I apologize in advance for slow replies on my part, I am traveling right now.)

**Comment by mccutchen** [comment]
> Also, have you explored what the original httpbin's behavior is in weird circumstances? Like, what happens with a request like `/response-headers?Content-Length=999999999`? I think we need some test coverage for at least that edge case, to codify whatever behavior we decide on.

The behavior is interesting (TLS-related noise elided below):

```
$ curl -v --http1.1 'https://httpbin.org/response-headers?Content-Type=text/plain&Content-Length=99999'
*   Trying 54.210.149.139:443...
* Connected to httpbin.org (54.210.149.139) port 443 (#0)
> GET /response-headers?Content-Type=text/plain&Content-Length=99999 HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/8.1.2
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Mon, 07 Aug 2023 22:40:24 GMT
< Content-Type: text/plain
< Content-Length: 99999
< Connection: keep-alive
< Server: gunicorn/19.9.0
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
{
  "Content-Length": [
    "122", 
    "99999"
  ], 
  "Content-Type": [
    "application/json", 
    "text/plain"
  ]
}
* transfer closed with 99877 bytes remaining to read
* Closing connection 0
curl: (18) transfer closed with 99877 bytes remaining to read
```

I can convince myself that letting the incoming request to this specific endpoint dictate the actual Content-Length of the response could be useful for testing weird behavior in HTTP client or proxy code, so I'd be in favor of matching the Python implementation there.

**Comment by mccutchen** [comment]
> How can I get just key - value?

I'm sorry, but I don't understand what you mean.  Can you provide a bit more context, or specific examples of what you'd like to see?

At the risk of over-explaining:  The HTTP protocol allows for both multiple headers and multiple URL query parameters with the same name.  This may be relatively rare, but it is entirely normal and compliant.  For go-httpbin — _unlike the original httpbin_ — we **do not** special-case the way we serialize single headers vs multiple headers or single query params vs multiple query params with a given name.  They are always serialized in go-httpbin's responses as arrays (single values as a 1-element array, multiple values as multiple-element arrays).

> Another way for this, use [] after key for list Ex:
> 
> ...
> 
> I know many frameworks use this way.

That is not part of the HTTP spec.  It is framework specific and entirely unnecessary, given a spec-compliant HTTP implementations.  I've never researched the history here, but I have always assumed that some frameworks made this implementation decision out of ignorance of the HTTP spec, but maybe there are historical reasons that necessitated this that I'm not aware of.

Regardless, go-httpbin is not going to implement any special cases like this.

(See https://github.com/mccutchen/go-httpbin/issues/37#issuecomment-770427446 for prior discussion of this `key[]=1&key[]=2` idea.)
